Add notes for October 2023 Hybrid meeting (#1436)
WebAssembly/meetings: main/2023/CG-10.md

CW: But would you ever try to import a less precise struct type by someone else or is that too much granularity

AR: Like a supertype? I don’t have a good use case for that. Maybe that exists in some systems but honestly I don’t know.

CW: That’s all I was hoping to hear

AR: Not sure if that adds any significant complexity. It’s just subtype checking and you can restrict the syntax on one side of that check or not, other than that I don’t think it has one real implication on implementations.

FM: You can’t compile an offset to a structure.. Maybe too specific

AR: What I meant is there is no implication for engines to allow this but disallowing that might be a nice example of excluding some things, knowing you have a common header for structs that define a different module without having to know the extent of all of them. That’s a good example. Might be worth just doing it, but Post-Post-MVP.

<AR Presenting>

TL: There’s one thorny but trivial problem, the type section comes before the import section

> AR: Yes, I should have that on the next slide. There is also one minor thing, we want to allow type definitions for that because someone might want to include i31 as their implementation for something.

Reviewer: ```suggestion
AR: Yes, I should have that on the next slide. There is also one minor thing, we want to allow type definitions with abstract heap types because someone might want to include i31 as their implementation for something.
```

<AR presenting Type Imports slide>

RT: For now when you cast to an imported type, so you can’t precompile the type during instantiation

AR: Will get back to that problem as well. You are right. We have to think about casts as well in that context.

<AR presenting Type Imports slide>

TL: Of course if you allow interleaving sections, that won't help the case where the import section has to come in the middle of a recursion group.

> AR: That won’t work anyway if you have mutual recursion, you can’t import anyway. That’s one thing about isorecursive types, you have to import them together

Reviewer: ```suggestion
AR: That won’t work anyway if you have mutual recursion, you can’t import them independently anyway. That’s one restriction about isorecursive types that I mentioned before, you have to import them together
```

<AR Presenting Type parameters>

TL: Are these so similar that every function that takes a type import is exactly the same thing? 

> AR: that’s the invariant I would want, a little module with an import that can access it’s surroundings so that it has a closure of course. Other than that, that’s the property I want. Efficient way that doesn’t require instantiation for each call.

Reviewer: ```suggestion
AR: that’s the invariant I would want, a generic function is equivalent to a little module with an import. It can access it’s surroundings so that it has a closure of course, other than that, that’s the property I want. Efficient way that doesn’t require instantiation for each call.
```

BT: I would like to see type parameters decoupled from type imports

> AR: No, as long as we keep the constraints separate

Reviewer: ```suggestion
AR: No need, as long as we keep the constraints separate. At least need to be designed together
```

BT: Seems like more engine work

> AR: that’s the point right this is pure validation time. Nothing in the interpreter and I think that’s an important goal to achieve. That’s why I want to keep it separate from value type instantiation.

Reviewer: ```suggestion
AR: that’s the point: right now this is a pure validation time feature. Needs nothing in the runtime and I think that’s an important goal to achieve. That’s why I want to keep it separate from value type instantiation.
```

FM: Before we go into that we should have a proper requirement or use case for it, which group of users are we unblocking? 

> AR: I think basically almost every language that has some form of generics except Java generics because they are an afterthought they are compiled away. Almost every other language that has generics has cases where if you don’t jhave that, you have to do this Uppercast. OCaml is one an example, C# is another.

Reviewer: ```suggestion
AR: I think basically almost every language that has some form of generics except Java generics because they are an afterthought they are compiled away. Almost every other language that has generics has cases where if you don’t have that, you have to do this upcast/downcast dance. OCaml is one an example, C# is another.
```

TL: Do the ocaml folks in the room have a sense of if they’ll use this?

PC: Given that we’ve seen max 10% in the test, we may not do it first

JK: Type related use case that stood out is would like to pass it back to the host and I don’t have to so far is importing opaque types from the host. Strings is one example but not the only. Would be good not to have a soup of externref because once you consume such an externref you need to type check it so if you don’t have something like Arrays and StringBuffers inside of Wasm core, and we have good reasons for some of these, it would be good to say I have a string from the host here and I don’t have to do a type check. Means we also need a type check function because the engine needs to do a type check at the boundary to amke sure it matches the object being passed around.

> AR: Which gets back to the question of moving back from exported types, another ay to phrase this imported/exported rtts..

Reviewer: ```suggestion
AR: Which gets back to the question of moving back from exported types, another way to phrase this imported/exported rtts..
```

CW: I'm a little surprised that you're expecting that we can map source level generics to this, given that we weren't expecting source types to map directly to Wasm types.

AR: They might have to do pass more than just an RTT. IT’s not sufficient, but it is a building block

CR: It’s a good start but I need to think more.

RT: For something like ocaml to use this, because it needs to downcast at some type. Every polymorphic function will also be gicen the rtt for it, you will then have to construct the rtt on the fly, and then you have to do the type construction dynamically, is this worth it? <TODO: RT to fix>

> AR: The whole point is not to pass an RTT everywhere. Basically when you compile something like OCaml with a type repr you have type_eq anyway. And only the places where you consume a value you need a cast but at that point you usually know the type you are casting to. In addition this allows a polymorphic call has a more concrete type, you might also get out a more concrete type when it gets back to you.

Reviewer: ```suggestion
AR: The whole point is not to pass an RTT everywhere. Basically when you compile something like OCaml with a type repr you have eqref anyway. And only the places where you consume a value you need a cast but at that point you usually know the type you are casting to. In addition this allows a polymorphic call has a more concrete type, you might also get out a more concrete type when it gets back to you.
```

RT: I can’t construct a pair value of alpha alpha, without knowing the rtt for tag (?)

> AR: That is correct. The other feature we need is a way of constructing/allocating values without RTT. So if you never want to cast them, they don’t need to carry RTT. A good question hwo much of your code would fall into the section of things where you never need RTT but it’s worth investigating whether it gives some benefit in circumstances. 

Reviewer: ```suggestion
AR: That is correct. The other feature we need is a way of constructing/allocating values without RTT. So if you never want to cast them, they don’t need to carry RTT. A good question hwo much of your code would fall into the intersection of things where you never need RTT but it’s worth investigating whether it gives some benefit in circumstances. 
```

ZB: Both features look like we would want to use them in Kotlin, more request - how about having import/export of unsigned numbers, could be used to provide better APIs and to go abc and forth from the host. 

AR: What do you mean importing unsigned numbers?

ZB: Allowing using u8 in import/export section

AR: Not just about imports/exports, we don’t have small integers. Once we have them you can import/export them as an immutable global but it is not something we have included.

ZB: Another thing about typed parameters is about covariant and  invariant things and different bounds for typed parameters

> AR: That’s a riddle. Something we could do is variance and polarities as they are sometimes called and plausible but not entirely straightforward and another added complexity. There is a point of diminishing returns but we have to figure out where that is, so for polarities we have to figure out where that is but nothing technically preventing that.

Reviewer: ```suggestion
AR: That’s a well-known problem. For that we could do variance and polarities as they are sometimes called. plausible but not entirely straightforward and another added complexity. There is a point of diminishing returns but we have to figure out where that is, so for polarities we have to figure out where that is but nothing technically preventing that.
```

BT: Scala has a long complicated history of type parameters, their experiments of reified ?? <BT to fix>

AR: maybe that’s why i'm bringing it up, worth investigating, I don’t have an answer. My gut is it is.

<we have no idea>: Just wanted to add what Ben keeps saying, I’m skeptical of type parameters at all because of the type varieties there are across the languages. 

> AR: most languages allow you to specify variance much, scala does, lets not talk about Java, doesn’t matter anyway. Some modern OOO languages have it, Ocaml doesn’t count

Reviewer: ```suggestion
AR: most languages don't allow you to specify variance much, scala does, lets not talk about Java, doesn’t matter anyway because type erasure. Some modern OOO languages have it, Ocaml doesn’t count
```

SL: ocaml does

AK: Whether we should do work here depends on source languages that can give us feedback. We shouldn’t sopend time without having folks singed up to do that.

> AR: They have to be able to experiment from it, chicken and egg problem

Reviewer: ```suggestion
AR: They have to be able to experiment with it, chicken and egg problem
```

AK: I think we should have people who say yes right now im hearing language implementers say I don’t know if I would try this because I have other stuff or its not a big deal. To answer the question about prioritization. 

AR: There’s a semantic difference for variance, if some languages use an unsound thing, they’re on their own

<AR Presenting Type parameters>

FM: So far this is really not post-MVP GC, but post-MVP types. What about extending the range of GC languages that we might want to support - Go, do we want to support finalizers?

> AR: I have a slide on finalizers yes. But yes this is mostly about types, but they are the complex part about GC mostly. If you look at a GC proposal mostly it’s trivial and most of the complexity is the types.

Reviewer: ```suggestion
AR: I have a slide on finalizers yes. But yes this is mostly about types, but they are the complex part about GC mostly. If you look at the GC proposal mostly it’s trivial and most of the complexity is the types.
```

> 

Reviewer: ```suggestion

<AR presenting Self types>

```

AR: Self Types you can read in two ways. It’s solving typing of self parameters

AK: That’s the right way to approach each of these things.

AR: Fair.

<AR Presenting Nested Aggregates>

FM: The java string type is actually an example of this.

AR: Right. Also in C when you do data structures in C, a common pattern to have an array at the end. Definitely not weird.

<AR Presenting Nested Aggregates>

TL: I don’t know if you’re familiar with the get Element pointer instruction in LLVM IR 

AR: I’m not

TL: It’s kind of like a struct.get, array.get but it takes as many indices as it needs to traverse into this nested thing, so you don't need to take references 

AR: Also not something you necessarily want over time, benefits for the complexity aren’t clear

CW: I remember early conversations about interior pointer support is this the same thing?

AR: But if engines implement fat pointers, they would be penalized, they’d have to optimize them into interiroir pointers. I doubt they would do that because interior pointers are painful. I have a slide saying I don’t know what we want to do here.

> 

Reviewer: ```suggestion

<AR presenting Finalizers>

```
